onmt/Trainer.py

```diff
@@ -0,0 +1,166 @@
+import torch
+import onmt
+import onmt.modules
+
+
+class Trainer(object):
+    def __init__(self, model, train_data, valid_data, fields, optim,
+                 batch_size, gpuid, copy_attn, copy_attn_force,
+                 truncated_decoder, max_generator_batches):
+        # Basic attributes.
+        self.model = model
+        self.train_data = train_data
+        self.valid_data = valid_data
+        self.fields = fields
+        self.optim = optim
+        self.batch_size = batch_size
+        self.gpuid = gpuid
+        self.copy_attn = copy_attn
+        self.truncated_decoder = truncated_decoder
+        self.max_generator_batches = max_generator_batches
+
+        # Define criterion.
+        padding_idx = fields['tgt'].vocab.stoi[onmt.IO.PAD_WORD]
```

Comment:
Remove all this copy stuff from this class. Ideally the trainer should not even know about NMT details at all.